### DISCUSSION TITLE
Pico Onewire and DS18B20

### DISCUSSION THREAD

**Discussion by rajivtctech** [earlephilhower/arduino-pico#389] [General]
I'm trying to read a DS18B20 temperature sensor with a RasPi Pico. Some libraries are obviously incompatible and compilation terminates with an error (e.g. a library function calling an AVR library). Other libraries compile with warnings and upload the program successfully, but you cannot then see your /dev/ttyACM0 port any longer... Why?

Has anyone read a Onewire DS18B20 temperature sensor successfully on the Pico, with the Arduino and Earle's core? If so, please point me to the library or method that worked for you...

RT

**Comment by Ing-Dom** [comment]
yes I did.
check this out, dude.
https://github.com/adamboardman/pico-onewire

**Reply by david-mcginnis** [reply]
Fantastic!

**Reply by hqdlab** [reply]
@Ing-Dom Is this really Arduino library? It looks more like code for sdk. What I and rajivtctech are looking for is an Arduino library for the Earle's RP2040 core.

**Reply by Ing-Dom** [reply]
I don't know what exactly you mean by a "Arduino library" but the linked code works with arduino-pico core as it includes the pico sdk.

**Reply by hqdlab** [reply]
What I mean by ''Arduino library" is a library such as the ones described here  https://support.arduino.cc/hc/en-us/articles/5145457742236-Add-libraries-to-Arduino-IDE# on how to install it. 
Your code does not look like Arduino code to me. It does not have the `void setup()` and the `void loop()` functions. 
Sorry to bother you and thanks for your help.

**Reply by Ing-Dom** [reply]
no, it is not an arduino library, but @rajivtctech did'nt ask for arduino lib, but for a solution that works within the arduino framework and with arduino-pico core.
That IS the case for the lib I linked.

**Reply by VeloSteve** [reply]
I have the same issue, and have to admit that I don't know how to use the adamboardman code in this context.  His docs refer to "your CMakeLists.txt", but when working with the Philhower pico-sdk in the Arduino IDE I don't see CMakeLists.txt.  

Let me ask it this way (maybe what the original poster meant, or not):  I can compile with the old #include <OneWire.h>
#include <DallasTemperature.h> lines, but the sensors are not recognized.  Is there a drop-in replacement which I can install which will allow the Arduino IDE to generate working code for the DS18B20?

If I have to switch to a new build system I will, but I get the feeling that we are talking about two different approaches here.

Thanks.

**Reply by Ing-Dom** [reply]
just copy the one_wire.cpp and one_wire.h to your project folder an use the funtions!

```
#include "one_wire.h"

int main() {
    stdio_init_all();
    One_wire one_wire(15); //GP15 - Pin 20 on Pi Pico
    one_wire.init();
    rom_address_t address{};
    while (true) {
        one_wire.single_device_read_rom(address);
        printf("Device Address: %02x%02x%02x%02x%02x%02x%02x%02x\n", address.rom[0], address.rom[1], address.rom[2], address.rom[3], address.rom[4], address.rom[5], address.rom[6], address.rom[7]);
        one_wire.convert_temperature(address, true, false);
        printf("Temperature: %3.1foC\n", one_wire.temperature(address));
        sleep_ms(1000);
    }
    return 0;
}
```

**Reply by VeloSteve** [reply]
Thank you very much!  This really helped.  I can now read a DS18B20 address and temperature.  My first attempt to read one of four sensors on the same pin failed, but I'm sure that's my mistake.

For other users of the Arduino IDE, note that you won't need the DallasTemperature.h file.  I tried using it at first, converting OneWire to One_wire, but DallasTemperature calls things which don't exist in the new library.  Better to use the approach demonstrated by Ing-Dom.  Here is my version of his code, converted to work in a more traditional Arduino way, and perhaps a bit more verbose:
```
#include "one_wire.h"

#define SENSOR_DATA 17
#define SENSOR_POWER 16

One_wire one_wire(SENSOR_DATA);
rom_address_t address{};

void setup() {
    pinMode(SENSOR_POWER, OUTPUT);
    digitalWrite(SENSOR_POWER, HIGH);
    Serial.begin(115200);
    //stdio_init_all();
    one_wire.init();
}

void loop() {
    one_wire.single_device_read_rom(address);
    Serial.print("Address: ");
    printAddress(address);
    Serial.println();
    one_wire.convert_temperature(address, true, false);
    Serial.print("Temperature: ");
    Serial.println(one_wire.temperature(address));
    //printf("Temperature: %3.1foC\n", one_wire.temperature(address));
    sleep_ms(1000);
}

void printAddress(rom_address_t &deviceAddress) {
  for (uint8_t i = 0; i < 8; i++)
  {
    // zero pad the address if necessary
    if (deviceAddress.rom[i] < 16) Serial.print("0");
    Serial.print(deviceAddress.rom[i], HEX);
  }
}
```
